src/Common/src/System/Net/HttpKnownHeaderNames.TryGetHeaderName.cs

```diff
@@ -12,191 +12,234 @@ internal static partial class HttpKnownHeaderNames
         /// Gets a known header name string from a matching char[] array segment, using an ordinal comparison.
         /// Used to avoid allocating new strings for known header names.
         /// </summary>
-        public static bool TryGetHeaderName(char[] key, int startIndex, int length, out string name)
+        public static bool TryGetHeaderName(char[] array, int startIndex, int length, out string name)
         {
-            CharArrayHelpers.DebugAssertArrayInputs(key, startIndex, length);
+            CharArrayHelpers.DebugAssertArrayInputs(array, startIndex, length);
+
+            return TryGetHeaderName(
+                array, startIndex, length,
+                (arr, index) => arr[index],
+                (known, a, start, len) => CharArrayHelpers.EqualsOrdinal(known, a, start, len),
+                out name);
+        }
+
+        /// <summary>
+        /// Gets a known header name string from a matching IntPtr buffer, using an ordinal comparison.
+        /// Used to avoid allocating new strings for known header names.
+        /// </summary>
+        public unsafe static bool TryGetHeaderName(IntPtr buffer, int length, out string name)
+        {
+            Debug.Assert(length >= 0);
+
+            if (buffer == IntPtr.Zero)
```

Comment:
Is length == 0 common?  I assume not, but if it is you could add the check here as well.